src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,589 @@
+
+    if (mStorage->SyncGetKeyValue(GetStorageKey(node).KeyName(), backingBuffer.Get(), len) != CHIP_NO_ERROR)
+    {
+        subscriptions.mSize = 0;
+        mStorage->SyncDeleteKeyValue(GetStorageKey(node).KeyName());
+        return CHIP_NO_ERROR;
+    }
+
+    TLV::ScopedBufferTLVReader reader(std::move(backingBuffer), len);
+
+    DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
+    TLV::TLVType arrayType;
+    DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.EnterContainer(arrayType));
+
+    size_t subscriptionCount;
+    DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.CountRemainingInContainer(&subscriptionCount));
+    if (subscriptionCount >= CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+    {
+        mStorage->SyncDeleteKeyValue(GetStorageKey(node).KeyName());
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    subscriptions.mSubscriptions =
+        std::unique_ptr<SubscriptionInfo[]>(new (std::nothrow) SubscriptionInfo[subscriptionCount + allocateExtraSpace]);
+    if (!subscriptions.mSubscriptions)
+    {
+        mStorage->SyncDeleteKeyValue(GetStorageKey(node).KeyName());
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    subscriptions.mSize = subscriptionCount;
+
+    for (size_t currentSubscriptionIndex = 0; currentSubscriptionIndex < subscriptionCount; currentSubscriptionIndex++)
+    {
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag()));
+
+        TLV::TLVType subscriptionContainerType;
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.EnterContainer(subscriptionContainerType));
+
+        subscriptions[currentSubscriptionIndex] = { .mNodeId = node.GetNodeId(), .mFabricIndex = node.GetFabricIndex() };
+
+        // Subscription ID
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(kSubscriptionIdTag));
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Get(subscriptions[currentSubscriptionIndex].mSubscriptionId));
+
+        // Min interval
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(kMinIntervalTag));
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Get(subscriptions[currentSubscriptionIndex].mMinInterval));
+
+        // Max interval
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(kMaxIntervalTag));
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Get(subscriptions[currentSubscriptionIndex].mMaxInterval));
+
+        // Fabric filtered boolean
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(kFabricFilteredTag));
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Get(subscriptions[currentSubscriptionIndex].mFabricFiltered));
+
+        // Attribute Paths
+        uint16_t pathCount = 0;
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Next(kPathCountTag));
+        DeleteSubscriptionsAndReturnLogErrorOnFailure(reader.Get(pathCount));
```

Comment:
We should be able to reuse code that TLV serializes attribute and event paths on the wire for read and subscribe requests.  Why are we reinventing the wheel for our persistence serialization format?

Reply:
At the very least, SubscribeRequestMessage does not currently contain Subscription ID.. but I see what you're saying - I can look into better code reuse before we ship.

Reply:
OK, I see the challenge.  One possibility could be to nest the existing subscription path TLV structure into a single entry in the top-level container, also containing the other needed fields.

I think I will just advocate that reuse would be nice if possible and defer to you on whether it is.  Either way, changing the structure in the future will obviously be more difficult, so we should try to have a structure now in the upcoming release that we like.

Reply:
Filed issue #24535